Reference Architecture (#933)
wandb/docs: docs/guides/hosting/self-managed/ref-arch.md

---
title: Reference Architecture
description: W&B Reference Architecture.
displayed_sidebar: default
---

> # Reference Architecture

Reviewer: ```suggestion
```
The page's title comes from the front matter and H1 shouldn't be specified in the Markdown.

> Weights & Biases provides a reference architecture outlining the recommended infrastructure and resources needed to support a deployment of the platform.

Reviewer: ```suggestion
This page describes a reference architecture for a Weights & Biases deployment and outlines the recommended infrastructure and resources to support a production deployment of the platform.
```

> Depending on your chosen deployment environment for Weights & Biases (W&B), various services are available to enhance the resiliency of your setup. For instance, major cloud providers offer robust managed database services, removing the complexity of configuring and maintaining database clustering and failover mechanisms. With these capabilities in mind, we’ve developed a reference architecture tailored to the most common deployment scenarios, ensuring seamless integration with cloud vendor services for optimal performance and reliability.

Reviewer: ```suggestion
Depending on your chosen deployment environment for Weights & Biases (W&B), various services can help to enhance the resiliency of your deployment.

For instance, major cloud providers offer robust managed database services, which remove the complexity of configuring and maintaining database clustering and failover mechanisms.

This reference architecture addresses some common deployment scenarios, ensuring seamless integration with cloud vendor services for optimal performance and reliability.
```

## Before you start

Consider carefully whether a self-managed approach with W&B is suitable for your team and specific requirements.

> Running any application in production comes with its own set of challenges, and W&B is no exception. While we aim to streamline the process, certain complexities may arise depending on your unique architecture and design decisions. Typically, managing a production deployment involves overseeing various components, including hardware, operating systems, networking, storage, security, the W&B platform itself, and other dependencies. This responsibility extends to both the initial setup of the environment and its ongoing maintenance.

Reviewer: ```suggestion
Running any application in production comes with its own set of challenges, and W&B is no exception. While we aim to streamline the process, certain complexities may arise depending on your unique architecture and design decisions. Typically, managing a production deployment involves overseeing various components, including hardware, operating systems, networking, storage, security, the W&B platform itself, and other dependencies. This responsibility extends to both the initial setup of the environment and its ongoing maintenance.

Consider carefully whether a self-managed approach with W&B is suitable for your team and specific requirements.
```
I suggest flipping these around.

> It is essential to have a strong understanding of running and maintaining production-grade applications if you choose to self-manage W&B. If your team needs assistance, our Professional Services team and partners offer support for implementation and optimization. For those who prefer a more managed and worry-free solution, we offer alternatives such as [W&B Multi-tenant Cloud](../hosting-options/saas_cloud.md) and [W&B Dedicated Cloud](../hosting-options/dedicated_cloud.md) deployment types.

Reviewer: ```suggestion
A strong understanding of how to run and maintain  production-grade application is an important prerequisite before you deploy self-managed W&B. If your team needs assistance, our Professional Services team and partners offer support for implementation and optimization.

For those who prefer a managed solution over a self-hosted deployment, we offer alternatives such as [W&B Multi-tenant Cloud](../hosting-options/saas_cloud.md) and [W&B Dedicated Cloud](../hosting-options/dedicated_cloud.md).
```
Do we really want to claim "worry-free"?

> ## Infrastructure diagram

Reviewer: ```suggestion
## Infrastructure
```

> ![](/images/hosting/reference_architecture.png)

Reviewer: ```suggestion
![W&B infrastructure diagram](/images/hosting/reference_architecture.png)
```

### Application layer

> The application layer consists of a Kubernetes cluster, which should include multiple nodes to ensure resilience against node failures. W&B runs as pods on the Kubernetes cluster.

Reviewer: ```suggestion
The application layer consists of a multi-node Kubernetes cluster, with resilience against node failures. The Kubernetes cluster runs and maintains W&B's pods.
```

### Storage layer

> The storage layer consists of a MySQL database and object storage. The MySQL database stores metadata and the object storage stores artifacts (models, datasets, and so on).

Reviewer: ```suggestion
The storage layer consists of a MySQL database and object storage. The MySQL database stores metadata and the object storage stores artifacts such as models and datasets.
```

## Infrastructure requirements

### Kubernetes

> W&B requires a Kubernetes cluster with a deployed, configured and fully functioning Ingress controller, along with the capability to provision Persistent Volumes. This is essential as the W&B server application comes as a [Kubernetes Operator](../operator.md) that deploys multiple Pods.

Reviewer: ```suggestion
The W&B server application is deployed as a [Kubernetes Operator](../operator.md) that deploys multiple Pods. For this reason, W&B requires a Kubernetes cluster with:
- A fully configured and functioning Ingress controller
- The capability to provision Persistent Volumes.
```

### MySQL

> W&B requires a MySQL database as a metadata store. The shape of the model parameters and related metadata impact the performance of the database. The database size grows as the ML practitioners track more training runs, and incurs read heavy load when queries are executed in run tables, users workspaces, and reports.

Reviewer: ```suggestion
W&B uses a MySQL database as a metadata store. The shape of the model parameters and related metadata impact the performance of the database. The database size grows as the ML practitioners track more training runs, and incurs read heavy load when queries are executed in run tables, users workspaces, and reports.
```

> Consider the following when you run your own MySQL database:

Reviewer: ```suggestion
Consider the following when you deploy a self-managed MySQL database:
```

1. **Backups**. You should periodically back up the database to a separate facility. W&B recommends daily backups with at least 1 week of retention.
2. **Performance.** The disk the server is running on should be fast. W&B recommends running the database on an SSD or accelerated NAS.
3. **Monitoring.** The database should be monitored for load. If CPU usage is sustained at > 40% of the system for more than 5 minutes it is likely a good indication the server is resource starved.

> 4. **Availability.** Depending on your availability and durability requirements you might want to configure a hot standby on a separate machine that streams all updates in realtime from the primary server and can be used to failover to in the event that the primary server crashes or become corrupted.

Reviewer: ```suggestion
- **Backups**. You should periodically back up the database to a separate facility. W&B recommends daily backups with at least 1 week of retention.
- **Performance.** The disk the server is running on should be fast. W&B recommends running the database on an SSD or accelerated NAS.
- **Monitoring.** The database should be monitored for load. If CPU usage is sustained at > 40% of the system for more than 5 minutes it is likely a good indication the server is resource starved.
- **Availability.** Depending on your availability and durability requirements you might want to configure a hot standby on a separate machine that streams all updates in realtime from the primary server and can be used to failover to in the event that the primary server crashes or become corrupted.
```

### Object storage

> W&B requires object storage (Amazon S3, Azure Cloud Storage, Google Cloud Storage, or any S3-compatible storage service) with Pre-signed URL and CORS support.

Reviewer: ```suggestion
W&B requires object storage with Pre-signed URL and CORS support, deployed in Amazon S3, Azure Cloud Storage, Google Cloud Storage, or a storage service compatible with Amazon S3.service) 
```

### Versions
* Kubernetes: at least version 1.29.
* MySQL: at least 8.0.

> 

Reviewer: I'd move these requirements to the Kubernetes and MySQL sections above (I think MySQL already has it)

> ## Other considerations

Reviewer: ```suggestion
```
Maybe we don't need the H2. WDYT?

### Networking

> In non-airgapped deployments, egress to the following endpoints during installation and during runtime is required:

Reviewer: ```suggestion
In a deployment connected a public or private network, egress to the following endpoints is required during installation _and_ during runtime:
```

    * deploy.wandb.ai
    * charts.wandb.ai

>     * docker.io

Reviewer: ```suggestion
    * `https://deploy.wandb.ai`
    * `https://charts.wandb.ai`
    * `https://docker.io`
```

    * quay.io

>     * gcr.io

Reviewer: ```suggestion
    * `https://quay.io`
    * `https://gcr.io`
```

> The training infrastructure as well as each desktop that tracks experiments needs access to W&B and to the object storage.

Reviewer: ```suggestion
Access to W&B and to the object storage is required for the training infrastructure and for each system that tracks the needs of experiments.
```

### DNS

> The fully qualified domain name of W&B should resolve to the IP address of the ingress/load balancer using an A record.

Reviewer: ```suggestion
The fully-qualified domain name (FQDN) of the W&B deployment must resolve to the IP address of the ingress/load balancer using an A record.
```

### SSL/TLS

> A valid, signed SSL/TLS certificate is required for secure communication between clients and W&B. SSL/TLS termination must be done on the ingress/load balancer. W&B server application does not terminate SSL/TLS.

Reviewer: ```suggestion
W&B requires a valid signed SSL/TLS certificate for secure communication between clients and the server. SSL/TLS termination must occur on the ingress/load balancer. The W&B server application does not terminate SSL or TLS connections.
```

> Please note: W&B does not recommend the use self-signed certificates and custom CAs.

Reviewer: Above way say that self-signed certs are not supported, here we say we don't recommend them. Which is it? :D

Reviewer 2: Not supported officially but configurable with some environment variable magic which is a nightmare to maintain. So we don't recommend customers any sort of self-signed/private CA signed certificates. Regardless, they bring those especially large enterprises don't use public CAs. So I'm a little torn on this one.

### Supported CPU architectures

> Only x86 architecture is supported.

Reviewer: ```suggestion
W&B runs on the Intel (x86) CPU architecture. ARM is not supported.
```

## Infrastructure provisioning

> The recommended way to deploy W&B for production is through the use of Terraform configuration that defines the required resources, their references to other resources and dependencies. W&B provides Terraform modules for the major clouds. Please refer to [Deploy W&B Server within self managed cloud accounts](../hosting-options/self-managed#deploy-wb-server-within-self-managed-cloud-accounts).

Reviewer: ```suggestion
Terraform is the recommended way to deploy W&B for production. Using Terraform, you define the required resources, their references to other resources, and their dependencies. W&B provides Terraform modules for the major clouds. For details, refer to [Deploy W&B Server within self managed cloud accounts](../hosting-options/self-managed#deploy-wb-server-within-self-managed-cloud-accounts).
```

## Sizing

> The tables below offer general guidelines to use as a starting point. W&B recommends monitoring all components closely and adjusting the sizing based on actual usage patterns.

Reviewer: ```suggestion
Use the following general guidelines as a starting point when planning a deployment. W&B recommends that you monitor all components of a new deployment closely and that you make adjustments based on observed usage patterns.
```

| ---------------- | ------------------ | ------------------ | ------------------ | 
| Test/Dev         | 4 cores            | 32 GB              | 100 GB             |
| Production       | 16 cores           | 128 GB             | 100 GB             |

Numbers are per Kubernetes worker node.

#### MySQL

| Environment      | CPU                | Memory             | Disk               | 
| ---------------- | ------------------ | ------------------ | ------------------ | 
| Test/Dev         | 2 cores            | 16 GB              | 100 GB             |
| Production       | 8 cores            | 64 GB              | 500 GB             |

Numbers are per MySQL node.

## Recommended cloud provider services and flavors

### Services

| Cloud       | Kubernetes	 | MySQL	                | Object Storage             |   
| ----------- | ------------ | ------------------------ | -------------------------- | 
| AWS         | EKS          | RDS Aurora               | S3                         |
| GCP         | GKE          | Google Cloud SQL - Mysql | Google Cloud Storage (GCS) |
| Azure       | AKS          | Azure Database for Mysql | Azure Blob Storage         |

> ### Flavors

Reviewer: I haven't seen this called "Flavors" but something like "machine types". Is this W&B terminology?

Reviewer 2: Nope, Machine types should be the right usage here. Agreed!

displayed_sidebar: default
---

This page describes a reference architecture for a Weights & Biases deployment and outlines the recommended infrastructure and resources to support a production deployment of the platform.

Depending on your chosen deployment environment for Weights & Biases (W&B), various services can help to enhance the resiliency of your deployment.

> For instance, major cloud providers offer robust managed database services, which remove the complexity of configuring and maintaining database clustering and failover mechanisms.

Reviewer: ```suggestion
For instance, major cloud providers offer robust managed database services which help to reduce the complexity of database configuration, maintenance, high availability, and resilience.
```

> This reference architecture addresses some common deployment scenarios, ensuring seamless integration with cloud vendor services for optimal performance and reliability.

Reviewer: ```suggestion
This reference architecture addresses some common deployment scenarios and shows how you can integrate your W&B deployment with cloud vendor services for optimal performance and reliability.
```
This aims to soften the promise a bit.

## Before you start

Running any application in production comes with its own set of challenges, and W&B is no exception. While we aim to streamline the process, certain complexities may arise depending on your unique architecture and design decisions. Typically, managing a production deployment involves overseeing various components, including hardware, operating systems, networking, storage, security, the W&B platform itself, and other dependencies. This responsibility extends to both the initial setup of the environment and its ongoing maintenance.

Consider carefully whether a self-managed approach with W&B is suitable for your team and specific requirements.

A strong understanding of how to run and maintain  production-grade application is an important prerequisite before you deploy self-managed W&B. If your team needs assistance, our Professional Services team and partners offer support for implementation and optimization.

> For those who prefer a managed solution over a self-hosted deployment, we offer alternatives such as [W&B Multi-tenant Cloud](../hosting-options/saas_cloud.md) and [W&B Dedicated Cloud](../hosting-options/dedicated_cloud.md).

Reviewer: ```suggestion
To learn more about managed solutions for running W&B instead of managing it yourself, refer to [W&B Multi-tenant Cloud](../hosting-options/saas_cloud.md) and [W&B Dedicated Cloud](../hosting-options/dedicated_cloud.md).
```

## Infrastructure

![W&B infrastructure diagram](/images/hosting/reference_architecture.png)

### Application layer

The application layer consists of a multi-node Kubernetes cluster, with resilience against node failures. The Kubernetes cluster runs and maintains W&B's pods.

### Storage layer

The storage layer consists of a MySQL database and object storage. The MySQL database stores metadata and the object storage stores artifacts such as models and datasets.

## Infrastructure requirements

### Kubernetes
The W&B Server application is deployed as a [Kubernetes Operator](../operator.md) that deploys multiple Pods. For this reason, W&B requires a Kubernetes cluster with:
- A fully configured and functioning Ingress controller
- The capability to provision Persistent Volumes.

### MySQL

> W&B uses a MySQL database as a metadata store. The shape of the model parameters and related metadata impact the performance of the database. The database size grows as the ML practitioners track more training runs, and incurs read heavy load when queries are executed in run tables, users workspaces, and reports.

Reviewer: ```suggestion
W&B stores metadata in a MySQL database. The database's performance and storage requirements depend on the shapes of the model parameters and related metadata. For example, the database grows in size as you track more training runs, and load on the database increases based on queries in run tables, users workspaces, and reports.
```

Consider the following when you deploy a self-managed MySQL database:

> - **Backups**. You should periodically back up the database to a separate facility. W&B recommends daily backups with at least 1 week of retention.
> - **Performance.** The disk the server is running on should be fast. W&B recommends running the database on an SSD or accelerated NAS.
> - **Monitoring.** The database should be monitored for load. If CPU usage is sustained at > 40% of the system for more than 5 minutes it is likely a good indication the server is resource starved.

Reviewer: ```suggestion
- **Backups**: You should periodically back up the database to a separate facility from the deployment. W&B recommends daily backups retained for a minimum of one week.
- **Performance**: The W&B server requires fast storage, such as solid-state disks (SSDs) or accelerated network attached storage (NAS).
- **Monitoring**: Monitor the database's load. If CPU utilization regularly remains above 40% for longer than 5 minutes on average, consider increasing CPU capacity on the server.
```
Some small reworking for active voice, brevity, clarity.

> - **Availability.** Depending on your availability and durability requirements you might want to configure a hot standby on a separate machine that streams all updates in realtime from the primary server and can be used to failover to in the event that the primary server crashes or become corrupted.

Reviewer: ```suggestion
- **Availability**: If you have additional requirements for availability or durability, consider configuring a "hot" standby W&B deployment on separate infrastructure. The active deployment can stream all updates in realtime to the standby server. If the primary deployment becomes unavailable, you can fail over to the standby deployment quickly.
```

### Object storage

> W&B requires object storage with Pre-signed URL and CORS support, deployed in Amazon S3, Azure Cloud Storage, Google Cloud Storage, or a storage service compatible with Amazon S3.service) 

Reviewer: ```suggestion
W&B requires object storage with support for Pre-signed URLs and CORS, deployed in Amazon S3, Azure Cloud Storage, Google Cloud Storage, or a storage service compatible with Amazon S3. 
```
A stray fragment at the end of the sentence, and some streamlining

> ### Versions

Reviewer: I suggest moving these version requirements up into the Kubernetes and MySQL sections above.

* Kubernetes: at least version 1.29.
* MySQL: at least 8.0.

### Networking

> In a deployment connected a public or private network, egress to the following endpoints is required during installation _and_ during runtime:

Reviewer: ```suggestion
In a deployment connected to a public or private network, egress to the following endpoints is required during installation _and_ during runtime:
```
What other kind of deployment do we envision? I guess a deployment connected to no networks. But ... is that possible? I suggest removing everything before "egress" and starting the sentence there.

    * `https://deploy.wandb.ai`
    * `https://charts.wandb.ai`
    * `https://docker.io`
    * `https://quay.io`
    * `https://gcr.io`

Access to W&B and to the object storage is required for the training infrastructure and for each system that tracks the needs of experiments.

### DNS

> The fully qualified domain name (FQDN) of the W&B deployment must resolve to the IP address of the ingress/load balancer using an A record.

Reviewer: ```suggestion
The fully qualified domain name (FQDN) of the W&B deployment must resolve to the IP address of the ingress/load balancer using an A record. A CN or other type of DNS record is not supported.
```
Is this what we mean?

### SSL/TLS

> W&B requires a valid signed SSL/TLS certificate for secure communication between clients and the server. SSL/TLS termination must occur on the ingress/load balancer. The W&B Server application does not terminate SSL or TLS connections.

Reviewer: ```suggestion
W&B requires a valid signed SSL/TLS certificate for secure communication between clients and the server. The ingress or load balancer handles the connection lifecycle. The W&B Server application does not manage SSL or TLS connections.
```
Still true? It's a bit more active and generalized.

> Please note: W&B does not recommend the use self-signed certificates and custom CAs.

Reviewer: ```suggestion
In a production deployment, W&B _strongly_ recommends against using self-signed certificates and custom CAs. If you require such a configuration, contact Support for guidance.
```
How is this?

### Supported CPU architectures
W&B runs on the Intel (x86) CPU architecture. ARM is not supported.

## Infrastructure provisioning

> Terraform is the recommended way to deploy W&B for production. Using Terraform, you define the required resources, their references to other resources, and their dependencies. W&B provides Terraform modules for the major clouds. For details, refer to [Deploy W&B Server within self managed cloud accounts](../hosting-options/self-managed#deploy-wb-server-within-self-managed-cloud-accounts).

Reviewer: ```suggestion
Terraform is the recommended way to deploy W&B for production. Using Terraform, you define the required resources, their references to other resources, and their dependencies. W&B provides Terraform modules for the major cloud providers. For details, refer to [Deploy W&B Server within self managed cloud accounts](../hosting-options/self-managed#deploy-wb-server-within-self-managed-cloud-accounts).
```

## Sizing

> Use the following general guidelines as a starting point when planning a deployment. W&B recommends that you monitor all components of a new deployment closely and that you make adjustments based on observed usage patterns.

Reviewer: ```suggestion
Use the following general guidelines as a starting point when planning a deployment. W&B recommends that you monitor all components of a new deployment closely and that you make adjustments based on observed usage patterns. Continue to monitor production deployments over time and make adjustments as needed to maintain optimal performance.
```
WDYT? Not one and done, probably.

Numbers are per MySQL node.

> ## Recommended cloud provider services and machine types

Reviewer: ```suggestion
## Cloud provider instance recommendations
```
Long headings are not great for SEO and don't look nice in the right-hand TOC. This is a bit more generic.

| Azure       | AKS          | Azure Database for Mysql | Azure Blob Storage         |

> ### Machine types

Reviewer: ```suggestion
### Machine types

These recommendations apply to each node of a self-managed deployment of W&B in cloud infrastructure.
```

| Test/Dev    | r6i.large          | r6i.xlarge         | r6i.xlarge          | db.r6g.large       | 
| Production  | r6i.2xlarge        | r6i.4xlarge        | r6i.4xlarge         | db.r6g.2xlarge     | 

> Machine type is per node.

Reviewer: If you take the recommendation above, you can remove this, 184, and 194.
